cmd/agent/main.go

```diff
@@ -70,6 +70,7 @@ type GrpcProxyAgentOptions struct {
 	adminServerPort  int
 
 	agentID       string
+	agentHost     string
```

Comment:
I generally like  this but wonder if we're short changing on the agentHost. I'm assuming this is just a single IP address and that we are running 1 agent per node. However we frequently give nodes IP ranges to assign to each pod running on that node its own individual IP address. (See https://cloud.google.com/kubernetes-engine/docs/concepts/network-overview). In addition we may have both IPv4 and IPv6 addresses for both nodes and pods. I wonder if it would be better to have a map of agent properties? Then we can easily extend this for things like exact IP match, range IP match, rang IPv6 match, failure zone match, etc. We could encode this as something like "host=foo,cidr=127.0.0.1/16,ipv6=:::::,failure-zone=us-central1-b". If we wanted to be really paranoid we could url encode it (https://en.wikipedia.org/wiki/Percent-encoding).

Reply:
I like the idea of encoding all potential addresses into a single string. What about something like the following,

1. when starting an agent, users can specify all potential addresses by passing a string, "host=foo,cidr=127.0.0.1/16,ipv6=:::::,failure-zone=us-central1-b" through the command line,

2. the agent will validate if the the addresses' string is valid and add it to the gRPC header,

3. then, the server will deserialize the addresses from the header, and add the agent to corresponding backendmanager.

As we can put the addresses' string into the request header(we are currently storing the `agentHost` and `agentID` in the gRPC stream header), do we really need to do URL encoding? I am not sure if the gRPC header can hold reserved characters of the Percent-encoding, but I will try it out.

Reply:
Question for both of you: How much of the information in `host=foo,cidr=127.0.0.1/16,ipv6=:::::,failure-zone=us-central1-b` etc is directly obtainable when starting the konnectivity agent pod? Are these parameters directly obtainable through something like environment variables or would additional logic be needed? Theoretically kube-apiserver would have this kind of information about the nodes, but adding additional logic there seems quite complicated. I'm wondering how we can support this in the simplest way.

Reply:
IMO, there may be two kinds of "agent address",

1. addresses that can be got directly when starting the konnectivity agent, like podIP(s), hostIP, etc,

2. addresses that added by user through cmd option, like, `--agent-addresses="host=foo,cidr=127.0.0.1/16,ipv6=:::::,failure-zone=us-central1-b"`, if some addresses are stored in env, we can pass them through the pod's template yaml. 

And it will be the users' (or whoever starts the agent ) responsiblity to ensure the provided addresses are reachable.

Reply:
There is also information which is probably category 1) but is not directly tied to the agent. So for instance things like failure zone are characteristics of the Node that the agent is running in. It can be looked up when the agent is started but its not clear to me that is when/how we will obtain that data. Its also possible it may be looked up by the (appropriate) backend manager when the agent is registered.